File: bulker/jitsubase/utils/clickhouse.go
package utils

import (
	"fmt"
	"net/url"
	"strings"
)

// ClickhouseConfig holds the parsed ClickHouse connection configuration
type ClickhouseConfig struct {
	Host     string
	Port     string
	Username string
	Password string
	Database string
	SSL      bool
}

// ClickhouseEnvVars interface for environment variable access
type ClickhouseEnvVars interface {
	GetClickhouseURL() string
	GetClickhouseHost() string
	GetClickhouseUsername() string
	GetClickhousePassword() string
	GetClickhouseDatabase() string
	GetClickhouseSSL() bool
}

// ParseClickhouseURL parses a ClickHouse URL and extracts components.
// URL format: [protocol://][username:password@]host[:port][/database]
func ParseClickhouseURL(rawURL string) (*ClickhouseConfig, error) {
	config := &ClickhouseConfig{}

	if rawURL == "" {
		return config, nil
	}

	// Add protocol if missing to make URL parsing work
	urlToParse := rawURL
	hasProtocol := strings.HasPrefix(rawURL, "http://") || strings.HasPrefix(rawURL, "https://")
	if !hasProtocol {
		// Check if it looks like it has credentials (contains @)
		if strings.Contains(rawURL, "@") {
			urlToParse = "https://" + rawURL
		} else {
			urlToParse = "https://" + rawURL
		}
	}

	parsed, err := url.Parse(urlToParse)
	if err != nil {
		return nil, fmt.Errorf("failed to parse ClickHouse URL: %w", err)
	}

	// Extract SSL from protocol
	if hasProtocol {
		config.SSL = parsed.Scheme == "https"
	}

	// Extract username and password
	if parsed.User != nil {
		config.Username = parsed.User.Username()
		config.Password, _ = parsed.User.Password()
	}

	// Extract host and port
	config.Host = parsed.Hostname()
	config.Port = parsed.Port()

	// Extract database from path
	if parsed.Path != "" && parsed.Path != "/" {
		config.Database = strings.TrimPrefix(parsed.Path, "/")
	}

	return config, nil
}

// GetClickhouseConfig builds ClickHouse configuration from environment variables.
// Priority:
// 1. Parse CLICKHOUSE_URL for all components
// 2. Fall back to individual env vars for missing components
// 3. Use CLICKHOUSE_SSL to determine SSL if not in URL
func GetClickhouseConfig(env ClickhouseEnvVars) (*ClickhouseConfig, error) {
	config := &ClickhouseConfig{}

	// Parse URL if provided
	clickhouseURL := env.GetClickhouseURL()
	if clickhouseURL != "" {
		parsed, err := ParseClickhouseURL(clickhouseURL)
		if err != nil {
			return nil, err
		}
		config = parsed
	}

	// Fall back to individual env vars for missing components
	if config.Username == "" {
		config.Username = NvlString(env.GetClickhouseUsername(), "default")
	}
	if config.Password == "" {
		config.Password = env.GetClickhousePassword()
	}
	if config.Database == "" {
		config.Database = env.GetClickhouseDatabase()
	}

	// Use CLICKHOUSE_SSL if SSL not determined from URL
	if clickhouseURL == "" || (!strings.HasPrefix(clickhouseURL, "http://") && !strings.HasPrefix(clickhouseURL, "https://")) {
		config.SSL = env.GetClickhouseSSL()
	}

	// Fall back to CLICKHOUSE_HOST if host not in URL
	if config.Host == "" {
		config.Host = env.GetClickhouseHost()
	}

	return config, nil
}

// GetAddr returns the address string for ClickHouse connection (host:port or just host)
func (c *ClickhouseConfig) GetAddr() string {
	if c.Port != "" {
		return c.Host + ":" + c.Port
	}
	return c.Host
}

// GetURL returns the full URL for ClickHouse HTTP connection
func (c *ClickhouseConfig) GetURL() string {
	scheme := "http"
	if c.SSL {
		scheme = "https"
	}
	return fmt.Sprintf("%s://%s", scheme, c.GetAddr())
}

File: libs/juava/src/clickhouse.ts
export type ClickhouseConfig = {
  url: string;
  username: string;
  password: string;
  database?: string;
};

export type ClickhouseEnvVars = {
  CLICKHOUSE_URL?: string;
  CLICKHOUSE_HOST?: string;
  CLICKHOUSE_USERNAME?: string;
  CLICKHOUSE_PASSWORD?: string;
  CLICKHOUSE_DATABASE?: string;
  CLICKHOUSE_METRICS_SCHEMA?: string;
  CLICKHOUSE_SSL?: string | boolean;
};

/**
 * Parses a ClickHouse URL and extracts components.
 * URL format: [protocol://][username:password@]host[:port][/database]
 */
export function parseClickhouseUrl(url: string): {
  protocol?: string;
  username?: string;
  password?: string;
  host?: string;
  port?: string;
  database?: string;
} {
  const result: ReturnType<typeof parseClickhouseUrl> = {};

  let remaining = url;

  // Extract protocol if present
  const protocolMatch = remaining.match(/^(https?):\/\//);
  if (protocolMatch) {
    result.protocol = protocolMatch[1];
    remaining = remaining.slice(protocolMatch[0].length);
  }

  // Extract username:password if present (before @)
  const atIndex = remaining.indexOf("@");
  if (atIndex !== -1) {
    const credentials = remaining.slice(0, atIndex);
    remaining = remaining.slice(atIndex + 1);

    const colonIndex = credentials.indexOf(":");
    if (colonIndex !== -1) {
      result.username = decodeURIComponent(credentials.slice(0, colonIndex));
      result.password = decodeURIComponent(credentials.slice(colonIndex + 1));
    } else {
      result.username = decodeURIComponent(credentials);
    }
  }

  // Extract database if present (after /)
  const slashIndex = remaining.indexOf("/");
  if (slashIndex !== -1) {
    result.database = remaining.slice(slashIndex + 1).split("?")[0] || undefined;
    remaining = remaining.slice(0, slashIndex);
  }

  // Remaining is host[:port]
  if (remaining) {
    const colonIndex = remaining.lastIndexOf(":");
    if (colonIndex !== -1) {
      result.host = remaining.slice(0, colonIndex);
      result.port = remaining.slice(colonIndex + 1);
    } else {
      result.host = remaining;
    }
  }

  return result;
}

/**
 * Builds ClickHouse client configuration from environment variables.
 *
 * Priority:
 * 1. Parse CLICKHOUSE_URL for all components
 * 2. Fall back to individual env vars for missing components
 * 3. Use CLICKHOUSE_SSL to determine protocol if not in URL
 *
 * @param env - Environment variables object
 * @returns ClickHouse client configuration
 */
export function getClickhouseConfig(env: ClickhouseEnvVars): ClickhouseConfig {
  let protocol: string | undefined;
  let username: string | undefined;
  let password: string | undefined;
  let host: string | undefined;
  let port: string | undefined;
  let database: string | undefined;

  // Parse URL if provided
  if (env.CLICKHOUSE_URL) {
    const parsed = parseClickhouseUrl(env.CLICKHOUSE_URL);
    protocol = parsed.protocol;
    username = parsed.username;
    password = parsed.password;
    host = parsed.host;
    port = parsed.port;
    database = parsed.database;
  }

  // Fall back to individual env vars for missing components
  username = username ?? env.CLICKHOUSE_USERNAME ?? "default";
  password = password ?? env.CLICKHOUSE_PASSWORD ?? "";
  database = database ?? env.CLICKHOUSE_METRICS_SCHEMA ?? env.CLICKHOUSE_DATABASE ?? "newjitsu_metrics";

  // Determine protocol from CLICKHOUSE_SSL if not in URL
  if (!protocol) {
    const ssl = env.CLICKHOUSE_SSL;
    const useSsl = ssl === true || ssl === "true" || ssl === "1";
    protocol = useSsl ? "https" : "http";
  }

  // Fall back to CLICKHOUSE_HOST if host not in URL
  if (!host) {
    host = env.CLICKHOUSE_HOST;
  }

  if (!host) {
    throw new Error("ClickHouse host is required. Set CLICKHOUSE_URL or CLICKHOUSE_HOST");
  }

  // Build the URL
  const portSuffix = port ? `:${port}` : "";
  const url = `${protocol}://${host}${portSuffix}`;

  return {
    url,
    username,
    password,
    database,
  };
}
